module/zfs/zfs_vfsops.c

```diff
@@ -972,6 +972,152 @@ zfs_id_overquota(zfsvfs_t *zfsvfs, uint64_t usedobj, uint64_t id)
 	    zfs_id_overobjquota(zfsvfs, usedobj, id));
 }
 
+static zfsvfs_kstats_t empty_zfsvfs_kstats = {
+	{ "writes",	KSTAT_DATA_UINT64 },
+	{ "nwritten",	KSTAT_DATA_UINT64 },
+	{ "reads",	KSTAT_DATA_UINT64 },
+	{ "nread",	KSTAT_DATA_UINT64 },
+	{ "dataset_name",	KSTAT_DATA_STRING },
+};
+
+static int
+zfsvfs_kstats_update(kstat_t *ksp, int rw)
+{
+	zfsvfs_t *zfsvfs = ksp->ks_private;
+	zfsvfs_kstats_t *zs = ksp->ks_data;
+
+	ASSERT3P(zfsvfs->z_iokstat->ks_data, ==, zs);
+
+	if (rw == KSTAT_WRITE)
+		return (EACCES);
+
+	zs->zk_writes.value.ui64 =
+	    aggsum_value(&zfsvfs->z_aggsum_stats.zas_writes);
+	zs->zk_nwritten.value.ui64 =
+	    aggsum_value(&zfsvfs->z_aggsum_stats.zas_nwritten);
+	zs->zk_reads.value.ui64 =
+	    aggsum_value(&zfsvfs->z_aggsum_stats.zas_reads);
+	zs->zk_nread.value.ui64 =
+	    aggsum_value(&zfsvfs->z_aggsum_stats.zas_nread);
+
+	return (0);
+}
+
+static void
+zfsvfs_kstats_create(zfsvfs_t *zfsvfs)
+{
+	ASSERT3P(zfsvfs->z_iokstat, ==, NULL);
+
+	/*
+	 * There should not be anything wrong with having kstats for
+	 * snapshots. Since we are not sure how useful they would be
+	 * though nor how much their memory overhead would matter in
+	 * a filesystem with many snapshots, we skip them for now.
+	 */
+	if (zfsvfs->z_issnap)
+		return;
+
+	/*
+	 * At the time of this writing, KSTAT_STRLEN is 255 in Linux,
+	 * the spa_name can be up to 256 characters and object IDs use
+	 * 48 bits maximum. Using the spa_name and the object ID (in
+	 * hex) separated by a dash should be good enough for most
+	 * pool names:
+	 * - pool name (1 to 256 chars)
+	 * - dash      (1 char)
+	 * - object id (1 to 12  chars in hex)
+	 *
+	 * Thus, pool names with lengths 1 to 242 should have valid
+	 * kstats created for their datasets. We do place an assertion
+	 * below to hopefully raise an issue when and if one of the
```

Comment:
If you could have a valid pool name that's 250 bytes long, failing an assertion in this case seems wrong.  A better behavior would be so silently not create the kstat (and maybe log a zfs_dbgmsg()).

However, the (current) maximum pool name length is 240 bytes:
```
int
pool_namecheck(const char *pool, namecheck_err_t *why, char *what)
{
	const char *c;

	/*
	 * Make sure the name is not too long.
	 * If we're creating a pool with version >= SPA_VERSION_DSL_SCRUB (v11)
	 * we need to account for additional space needed by the origin ds which
	 * will also be snapshotted: "poolname"+"/"+"$ORIGIN"+"@"+"$ORIGIN".
	 * Play it safe and enforce this limit even if the pool version is < 11
	 * so it can be upgraded without issues.
	 */
	if (strlen(pool) >= (ZFS_MAX_DATASET_NAME_LEN - 2 -
	    strlen(ORIGIN_DIR_NAME) * 2)) {
		if (why)
			*why = NAME_ERR_TOOLONG;
		return (-1);
	}
```

So the code you have is probably good, but we should update the comment to explain why this works (i.e. pool name is up to 240 chars).

EXCEPT, the above length analysis looks wrong because it doesn't account for the length of the literal `"/objset-0x"` in the name (it only counts 1 byte for the `-`).